src/mame/sinclair/specnext_divmmc.cpp

```diff
@@ -0,0 +1,173 @@
+// license:BSD-3-Clause
+// copyright-holders:Andrei I. Holub
+/**********************************************************************
+
+  Spectrum Next DivMMC
+
+**********************************************************************/
+
+#include "emu.h"
+#include "specnext_divmmc.h"
+
+
+specnext_divmmc_device::specnext_divmmc_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, SPECNEXT_DIVMMC, tag, owner, clock)
+{
+}
+
+void specnext_divmmc_device::cpu_mreq_n_w(bool data)
+{
+	m_cpu_mreq_n = data;
+	if (m_cpu_mreq_n)
+		automap_held();
+}
+
+void specnext_divmmc_device::cpu_m1_n_w(bool data)
+{
+	m_cpu_m1_n = data;
+}
+
+void specnext_divmmc_device::automap_active_w(bool data)
+{
+	m_automap_active = data;
+	automap_hold();
+}
+
+void specnext_divmmc_device::automap_rom3_active_w(bool data)
+{
+	m_automap_rom3_active = data;
+	automap_hold();
+}
+
+bool specnext_divmmc_device::conmem()
+{
+	return BIT(m_divmmc_reg, 7);
+}
+
+bool specnext_divmmc_device::mapram()
+{
+	return BIT(m_divmmc_reg, 6);
+}
+
+bool specnext_divmmc_device::page0()
+{
+	return m_cpu_a_15_13 == 0;
+}
+
+bool specnext_divmmc_device::page1()
+{
+	return m_cpu_a_15_13 == 1;
+}
+
+bool specnext_divmmc_device::rom_en()
+{
+	return page0() && (conmem() || automap()) && !mapram();
+}
+
+bool specnext_divmmc_device::ram_en()
+{
+	return (page0() && (conmem() || automap()) && mapram()) || (page1() && (conmem() || automap()));
+}
+
+u8 specnext_divmmc_device::ram_bank()
+{
+	return page0() ? 3 : (m_divmmc_reg & 0x0f);
+}
+
+bool specnext_divmmc_device::button_nmi()
+{
+	if (m_automap_reset || m_retn_seen)
+		m_button_nmi = 0;
+	else if (m_divmmc_button)
+		m_button_nmi = 1;
+	else if (automap_held())
+		m_button_nmi = 0;
+
+	return m_button_nmi;
+}
+
+bool specnext_divmmc_device::automap_nmi_instant_on()
+{
+	return m_automap_nmi_instant_on && button_nmi();
+}
+
+bool specnext_divmmc_device::automap_nmi_delayed_on()
+{
+	return m_automap_nmi_delayed_on && button_nmi();
+}
+
+bool specnext_divmmc_device::automap_hold()
+{
+	if (m_automap_reset || m_retn_seen)
+		m_automap_hold = 0;
+	else if (!m_cpu_mreq_n && !m_cpu_m1_n)
+		m_automap_hold = (m_automap_active && (m_automap_instant_on || m_automap_delayed_on || automap_nmi_instant_on() || automap_nmi_delayed_on()))
+			|| (m_automap_rom3_active && (m_automap_rom3_instant_on || m_automap_rom3_delayed_on))
+			|| (automap_held() && !(m_automap_active && m_automap_delayed_off));
+
+	return m_automap_hold;
+}
+
+bool specnext_divmmc_device::automap_held()
+{
+	if (m_automap_reset || m_retn_seen)
+		m_automap_held = 0;
+	else if (m_cpu_mreq_n)
+		m_automap_held = automap_hold();
+
+	return m_automap_held;
+}
```

Comment:
This is called from `automap_held_r``, which is named like a read handler.  Can it be called (directly or indirectly) from an address space read handler?  If it can, you need to check machine().side_effects_disabled() before changing the state.

Reply:
This can still be called from `divmmc_rom_en_r()` and `divmmc_ram_en_r()` (via `rom_en()` and `ram_en()`, respectively, which both call `automap()` which then calls `automap_held()`, which are named like read handlers.

I briefly looked at the code in the other PR, and it seems they’re only called from `specnext_state::bank_update`, which shouldn’t be called where side effects aren’t expected.

Should `divmmc_rom_en_r()` and `divmmc_ram_en_r()` be renamed to make it clearer they that they aren’t just getting signal states and have side effects?  Can they be called `try_rom_en()` and `try_ram_en()`, or `request_rom_en()` and `request_ram_en()`, or something?

Reply:
Yeah. Let me reflect on this.
If that's the only thing you unhappy about here, fill free to merge it. I'll catch up on this in the main PR.

Reply:
Must be cleaner this way